ambry-account/src/main/java/com/github/ambry/account/CachedAccountService.java

```diff
@@ -0,0 +1,652 @@
+            AccountServiceErrorCode.ResourceConflict);
+      }
+      for (Account account : accounts) {
+        if (accountInfoMap.hasConflictingContainer(account.getAllContainers(), account.getId(),
+            config.ignoreVersionMismatch)) {
+          logger.error(
+              "Containers={} under Account={} conflict with the containers in local cache. Cancel the update operation.",
+              account.getAllContainers(), account.getId());
+          accountServiceMetrics.updateAccountErrorCount.inc();
+          throw new AccountServiceException("Containers in account " + account.getId() + " conflict with local cache",
+              AccountServiceErrorCode.ResourceConflict);
+        }
+      }
+    } finally {
+      infoMapLock.readLock().unlock();
+    }
+
+    // write added/modified accounts to database
+    try {
+      updateAccountsWithMySqlStore(accounts);
+    } catch (SQLException e) {
+      logger.error("Failed updating accounts={} in MySql DB", accounts, e);
+      handleSQLException(e);
+    }
+    // Tell the world
+    publishChangeNotice();
+
+    if (writeCacheAfterUpdate) {
+      // Write accounts to in-memory cache. Snapshot version will be out of date until the next DB refresh.
+      updateAccountsInCache(accounts);
+    }
+  }
+
+  @Override
+  public void updateResolvedContainers(Account account, Collection<Container> resolvedContainers)
+      throws AccountServiceException {
+    try {
+      updateContainersWithMySqlStore(account, resolvedContainers);
+    } catch (SQLException e) {
+      logger.error("Failed updating containers {} in MySql DB", resolvedContainers, e);
+      handleSQLException(e);
+    }
+    // Tell the world
+    publishChangeNotice();
+
+    if (writeCacheAfterUpdate) {
+      // Write containers to in-memory cache. Snapshot version will be out of date until the next DB refresh.
+      updateContainersInCache(resolvedContainers);
+    }
+  }
+
+  @Override
+  public Collection<Container> updateContainers(String accountName, Collection<Container> containers)
+      throws AccountServiceException {
+    if (mySqlAccountStore == null) {
+      mySqlAccountStore = supplier.get();
+    }
+    try {
+      return super.updateContainers(accountName, containers);
+    } catch (AccountServiceException ase) {
+      if (needRefresh) {
+        // refresh and retry (with regenerated containerIds)
+        try {
+          fetchAndUpdateCache();
+        } catch (SQLException e) {
+          throw translateSQLException(e);
+        }
+        accountServiceMetrics.conflictRetryCount.inc();
+        return super.updateContainers(accountName, containers);
+      } else {
+        throw ase;
+      }
+    }
+  }
+
+  @Override
+  protected void publishChangeNotice() {
+    // TODO: can optimize this by sending the account/container payload as the topic message,
+    // so subscribers can update their cache and avoid the database query.
+    super.publishChangeNotice();
+  }
```

Comment:
Minor: We probably don't need to override the base method since there is no additional functionality in derived method.

Reply:
I just want to keep the todo item. Not quite sure why previously we have the todo in MysqlAccountService, but if you think we don't need that, I can remove in follow up pr.